File: src/CannabisMath.Core/Packaging/PackageCalculator.cs
namespace CannabisMath.Core.Packaging;

public static class PackageCalculator
{
    public static int FullUnits(decimal totalGrams, decimal gramsPerUnit)
    {
        ValidateNonNegative(totalGrams, nameof(totalGrams));
        ValidatePositive(gramsPerUnit, nameof(gramsPerUnit));

        return (int)(totalGrams / gramsPerUnit);
    }

    public static decimal RemainingGrams(decimal totalGrams, decimal gramsPerUnit)
    {
        ValidateNonNegative(totalGrams, nameof(totalGrams));
        ValidatePositive(gramsPerUnit, nameof(gramsPerUnit));

        var fullUnits = FullUnits(totalGrams, gramsPerUnit);
        return totalGrams - (fullUnits * gramsPerUnit);
    }

    public static (int fullUnits, decimal remainingGrams) Breakdown(decimal totalGrams, decimal gramsPerUnit)
    {
        var units = FullUnits(totalGrams, gramsPerUnit);
        var remainder = RemainingGrams(totalGrams, gramsPerUnit);

        return (units, remainder);
    }

    private static void ValidateNonNegative(decimal value, string parameterName)
    {
        if (value < 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "Value cannot be negative.");
        }
    }

    private static void ValidatePositive(decimal value, string parameterName)
    {
        if (value <= 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "Value must be greater than zero.");
        }
    }

    public static PackageBreakdownResult CalculateBreakdownResult(decimal totalGrams, decimal gramsPerUnit)
    {
        var fullUnits = FullUnits(totalGrams, gramsPerUnit);
        var remainingGrams = RemainingGrams(totalGrams, gramsPerUnit);

        return new PackageBreakdownResult(
            TotalGrams: totalGrams,
            GramsPerUnit: gramsPerUnit,
            FullUnits: fullUnits,
            RemainingGrams: remainingGrams
        );
    }
}

File: js/src/packaging/packageCalculator.ts
import type { PackageBreakdownResult } from "./packageBreakdownResult";

export class PackageCalculator {
  static fullUnits(totalGrams: number, gramsPerUnit: number): number {
    this.validateNonNegative(totalGrams, "totalGrams");
    this.validatePositive(gramsPerUnit, "gramsPerUnit");

    return Math.floor(totalGrams / gramsPerUnit);
  }

  static remainingGrams(totalGrams: number, gramsPerUnit: number): number {
    this.validateNonNegative(totalGrams, "totalGrams");
    this.validatePositive(gramsPerUnit, "gramsPerUnit");

    const remainder = totalGrams % gramsPerUnit;
    return this.round(remainder, 10);
  }

  static calculateBreakdownResult(
    totalGrams: number,
    gramsPerUnit: number
  ): PackageBreakdownResult {
    const fullUnits = this.fullUnits(totalGrams, gramsPerUnit);
    const remainingGrams = this.remainingGrams(totalGrams, gramsPerUnit);

    return {
      totalGrams,
      gramsPerUnit,
      fullUnits,
      remainingGrams,
    };
  }

  private static round(value: number, decimals: number): number {
    const factor = 10 ** decimals;
    return Math.round(value * factor) / factor;
  }

  private static validateNonNegative(value: number, name: string): void {
    if (value < 0) {
      throw new Error(`${name} cannot be negative`);
    }
  }

  private static validatePositive(value: number, name: string): void {
    if (value <= 0) {
      throw new Error(`${name} must be greater than zero`);
    }
  }
}